goodtables/compat.py

```diff
@@ -30,22 +31,22 @@
     basestring = basestring
     numeric_types = (int, long, float)
 
-    def csv_reader(data, dialect=csv.excel, **kwargs):
+    def csv_reader(data, dialect=csv.excel, header_index=0, **kwargs):
         """Read text stream (unicode on Py2.7) as CSV."""
 
-        first_lines = list(islice(data, 5))
-        try:
-            dialect = csv.Sniffer().sniff(''.join(first_lines))
-            dialect.delimiter = dialect.delimiter.encode('utf-8')
-            dialect.quotechar = dialect.quotechar.encode('utf-8')
-        except csv.Error:
-            dialect = csv.excel
-
+        first_lines = list(islice(data, header_index, header_index + 2))
+        dialect = detect_dialect(first_lines)
+        
+        while not re.findall('[^\w ]', dialect.delimiter): 
+            first_lines.extend(list(data.readline()))
+            dialect = detect_dialect(first_lines)
+            
         def iterenc_utf8(data):
             for line in data:
                 yield line.encode('utf-8')
 
-        iter = chain(first_lines, data)
+        data.seek(0)
```

Comment:
Adam, functions in DataTable (and maybe other places) expect to start from 0: https://github.com/frictionlessdata/goodtables/blob/master/goodtables/datatable/datatable.py#L209. I started sampling from the `header_index` because it influences the accuracy of the detection but if I then use `chain`, it will only include the lines from `header_index` downwards which conflicts with what other functions expect and leads to incorrect results.